### ISSUE TITLE
SFML 2.3 fullscreen bug Linux

### ISSUE THREAD

**Issue by matus123** [SFML/SFML#921]
hello, i have encountered this bug with fullscreen window in SFML 2.3, when i create window with default desktop size it sometimes creates window with correct size and viewSize , but more often it creates fullscreen window with broken size 

example:
normale window size 1366 - 768
broken window size 1366 - 714 with same viewSize
1920 - 1080
1920 - 1026

sample code : https://gist.github.com/matus123/967f48ea849fb670cf13

it behaves the same way even with correct VideoMode or with wrong as you can see in example

i have tried this code even with SFML 2.2 and it behaves the same way.

i am using Linux mint KDE

does anyone encountered similar problem with fullscreen window?, can anyone help me fix it? 

i will try this code on other linux distros to see if it's global or it is only misbehaving on my machine.

**Comment by dabbertorres** [comment]
Based on the size change only in the vertical component, I'm going to
assume it's taking your taskbar size into account sometimes, and sometimes
not.

Can you determine if you have anything running in the background that
causes the difference in behavior?

I can try this later today, but I haven't noticed anything like this.

Though, this will probably be closed and moved to the forums.

**Comment by matus123** [comment]
I tried changing taskbar visibility to auto hide and now it shows 1366 - 746.
So it seems that you had been right about taskbar. But anyway, i don't get why it is sometimes taking taskbar into the account and sometimes not.

I tried to run it from terminal right after boot with hidden taskbar.

I will test it on another linux distro with different window manager to see if the problem occurs.

**Comment by zryan3** [comment]
I have no issues going full-screen with Ubuntu 14.04 (Unity desktop).

**Comment by dabbertorres** [comment]
Okay, hmm. I guess I haven't tried fullscreen in a while. 

I'm getting "Failed to set new screen configuration" in stdout, which comes [from here](https://github.com/SFML/SFML/blob/26fc87284439699a720fe4b70563d3ebbd0b50bd/src/SFML/Window/Unix/WindowImplX11.cpp#L1464).

Seems to be failing to make the window fullscreen, and is making the window a borderless "fullscreen" window.
I feel like I should be getting "Failed to find matching fullscreen size, switching to window mode" as well, given that it's making a normal window, and not a "normal" fullscreen window, but I do not receive that message, which makes sense, looking at the code there. 

[Looks like](http://xcb.freedesktop.org/manual/structxcb__generic__error__t.html) there could be more information given about the error - I should have some time this weekend to add some code to get that information, and recompile SFML and see what I get.

Anyway, system:
Arch Linux x86_64
Kernel: 4.0.7-2-ARCH
DE: Cinnamon 2.6.12

**Comment by matus123** [comment]
This is really interesting, because i tried it on another machine with newly installed Linux Mint 17.2 MATE, i have also built SFML-2.3 on that machine.

I have encountered very similar results as before on my main PC with Linux Mint 17.1 KDE, which are based on Ubuntu 14.04

Screenshot of results : http://postimg.org/image/jx4x21y5p/

**Comment by dabbertorres** [comment]
Funny, you're getting a different error than I am. 

You should try something like [this](https://gist.github.com/dabbertorres/b18512f04fa795a30282), and see what the output is, and compare to what you VideoMode(s) you're trying.

**Comment by matus123** [comment]
I tried to print my supported modes. Also i tried to create window with my DefaultDesktopMode() but the result was the same. Sometimes it was correct size and sometimes modified.

I have also tried creating window with unsupported mode, because if i am correct sfml will firstly check if it's valid VideoMode and if not it will replace it with sf::VideoMode::getFullscreenModes()[0].

**Comment by TankOs** [comment]
Do you guys use multiple monitors? Can you paste the output of `xrandr -q`?

**Comment by matus123** [comment]
i am using notebook with connected monitor, but i am only using monitor display.
First image is with my main pc/notebook.
![hodnoty1](https://cloud.githubusercontent.com/assets/7724533/9195543/e2efd5f8-4023-11e5-877f-f534b43f17ad.png)

This image was taken on an old machine which was connected to TV via VGA cable.
![screenshot-1](https://cloud.githubusercontent.com/assets/7724533/9195551/ec3978ee-4023-11e5-865a-98cce83fe503.png)

**Comment by binary1248** [comment]
Does this still occur with the current master revision of SFML?

**Comment by bloodsword** [comment]
Hello,
I got this bug with the current master revision on 3 tested KDE desktop (Debian & Fedora).
I dug into SFML code, it seems that xcb is returning a wrong size with xcb_get_geometry in WindowImplX11::getSize().
I found two solutions to get rid of this bug by modifying WindowImplX11.xpp :
- force fullscreen at window creation by not testing ewmhSupported.  
OR
- apply the "OpenBox" workaround for "KWin".

I don't know the possible side effects of the first solution, but the second solution seems to be better because there is already a fix for OpenBox, but according to matus123 it also happens on MATE, 
so maybe it would be better to apply the workaround no matter the windows manager since it looks harmless for other managers.

**Comment by Eremiell** [comment]
Tried to mess with this one a bit today as it's still an issue. I'm on MATE btw.

Some [screenshots](http://imgur.com/a/GrA6r) to be referred. The white rectangle is of expected window size.

Some observations:

* with both 2.3.2 and latest commit all of the screenshots (figures 1-3) happen. It's always the first one most time with some amount of second or third time, depending on video mode.
* both 2.3.1 and 2.3 do rarely the first screen and go alright fullscreen about 80% of time
* 2.2 goes completely broken (the rectangle is drawn in the figure 1 position over the current display, mouse pointer can be moved but can't leave the rectangle and the whole X session doesn't react to anything, have to kill the test from second virtual terminal, so no screenshots possible), didn't go any deeper
* sleeping or recreating the window doesn't help

Output of the first program: (modified the 1366 to 1360 as that's what my screen supports)
```
window-view-size 1360 768
window-size 1920 768
```

[Output of the second program](http://hastebin.com/uguvewicun.md)

Tried to "apply the OpenBox workaround" (not sure if properly), mostly messed with [this](https://github.com/SFML/SFML/blob/94dd1d7/src/SFML/Window/Unix/WindowImplX11.cpp#L502) line, playing with the conditions, it either didn't change anything from "normal" state, or forced right size, but remained windowed.

To add some more confusion, on 2.3.1, the first tool sometimes returns the same as on 2.3.2 and sometimes this:
```
window-view-size 1920 1080
window-size 1920 1080
```
which looks like, it's not really going into 1360x768 as expected, but stays 1920x1080, but that actually seems to be the error state, as it's happening very rare compared to the former one and when running my test code, I get working behaviour most of the time and broken rarely.

Btw, here's my [test code](http://hastebin.com/ekikufurox.cpp) and the test image is the last one in the imgur album. (The screenshots were made with borderless version, I only added border once I got to 2.3.1 and the borders became relevant.)

So it seems, something got more wrong since [2.3.1](https://github.com/SFML/SFML/compare/2.3.1...2.3.2), but even there, it's not 100% success. Seems to me only event and input stuff got changed there, but maybe someone can see something I can't.

**Comment by dubmosphere** [comment]
Hello,
I have a similar problem on Linux Mint 17 Cinnamon. It only occurs with my tilemap drawables taking sf::Drawable and sf::Transformable as base class. The code works on windows, so this is really confusing. If I go to 800x600 the view gets scaled but not the drawables, they are drawed at the bottom left of the view.

**Comment by achpile** [comment]
> Does this still occur with the current master revision of SFML?

Yes, still occur =(

**Comment by texus** [comment]
I'm having this issue as well on Arch Linux with Cinnamon and the latest SFML version. Sometimes it works fine and the size is 1366x768 as expected but sometimes it incorrectly prints 1364x716.

In the cases where the size reported by window.getSize() and window.getView().getSize() is incorrect, the contents of the window is shifted, but the mouse coordinates in the sf::Event::MouseMoved event are still as if the window was 1366x768.

Applying the "OpenBox workaround" seems to work for me, changing [that line in WindowImplX11.cpp](https://github.com/SFML/SFML/blob/master/src/SFML/Window/Unix/WindowImplX11.cpp#L504) to `if (!(style & Style::Resize))` solved the problem here.

**Comment by IsemanTech** [comment]
I can confirm 
OS: Arch Linux
DE: KDE Plasma

SFML spits out:
`The requested video mode is not available, switching to a valid mode` 
I'll checkout the code in the fullscreen function

**Comment by eXpl0it3r** [comment]
Make sure to check out the bugfix/xlib branch (see PR #1138 and [builds for the lazy](http://www.sfml-dev.org/artifacts/by-branch/bugfix/xlib/)) and see if this solves the issue!

**Comment by texus** [comment]
With the bugfix/xlib branch merged the view size and window size are correct (1366x768) while SFML 2.4.0 reported 1364x714 and 1366x714 so it does seem to fix the issue.

**Comment by eXpl0it3r** [comment]
Should be fixed through #1138 in 1dc3db030818297c6ebeae43c85eb3e28b051a07.
